dlr_vieten/explorer/templates/dlr_vieten_explorer_tolerance_factors.html

```diff
@@ -25,17 +26,202 @@
+                        subelement.addClass('form-control').prop('type', 'text')
+                            .prop('placeholder', data[c][first_key]);
+                        element.append(subelement);
+                        var subelement = $('<span>');
+                        subelement.addClass('input-group-addon').html(first_key);
+                        element.append(subelement);
+                        wrapper.append(element);
+                        line.append(wrapper);
+                    }
+                };
+                $('#listTM').append(line);
+            });
+            $('#go1').on('click', function (e) {
+                var t = parseFloat($('#t').attr('placeholder'));
+                var r = parseFloat($('#r').attr('placeholder'));
+                var A_avg_Mg = 0; var A_avg_Ca = 0; var A_avg_Sr = 0; var A_avg_Ba = 0; var rO;
+                for (i = 0; i < table['rows'].length; i++) {
+                    var row = table['rows'][i];
+                    if (row['type'] == 'alkali earth') {
+                        var r_ion = parseFloat(row['rᵢₒₙ'].split()[0]);
+                        if (row['CN'] == '12') { if (row ['species'] == 'Mg')  { A_avg_Mg = r_ion * r; }}
+                        if (row['CN'] == '10') { if (row ['species'] == 'Mg')  { A_avg_Mg += r_ion * (1-r); }}
+                        if (row['CN'] == '12') { if (row ['species'] == 'Ca')  { A_avg_Ca = r_ion * r; }}
+                        if (row['CN'] == '10') { if (row ['species'] == 'Ca')  { A_avg_Ca += r_ion * (1-r); }}
+                        if (row['CN'] == '12') { if (row ['species'] == 'Sr')  { A_avg_Sr = r_ion * r; }}
+                        if (row['CN'] == '10') { if (row ['species'] == 'Sr')  { A_avg_Sr += r_ion * (1-r); }}
+                        if (row['CN'] == '12') { if (row ['species'] == 'Ba')  { A_avg_Ba = r_ion * r; }}
+                        if (row['CN'] == '10') { if (row ['species'] == 'Ba')  { A_avg_Ba += r_ion * (1-r); }}
+                    } else if (row['type'] == 'oxide ion') {
+                        rO = parseFloat(row['rᵢₒₙ'].split()[0]);
+                    }
+                }
+
+                //var nrTMs = $('#listTM').children.length;
+                var TMs = (1./nrTMs) // TODO  //
+                var TM3 = []; var TM4 = [];
+                $('#listTM').find('input').each(function (index) {
+                    var r_ion = parseFloat($(this).attr('placeholder').split()[0]);
+                    if (1) { TM4.push(r_ion); } //index%2 removed for now
+                    else { TM3.push(r_ion); }
+                });
+                var TM_avg = 0;
+                for (i = 1; i <= nrTMs; i++) { TM_avg += TM4[(3*i)-1] * TMs; }  //replace TMs by TM4[(3*i)-3] to get actual stoichiometries (when placeholders are removed)
+                var A_targ = Math.SQRT2 * t * (TM_avg + rO) - rO;
+                var errormsg = {};
+                var Mg = 0; var Ca = 0; var Sr = 0; var Ba = 0;
```

Comment:
This part should also be handled using a dictionary to avoid code duplication.

Reply:
this code has changed entirely now, no more duplication here